[Jetpack focus] Remove carousel and implement WordPress landing screen redesign

mkevins opened the pull request:
Fixes #17072

### Description

This PR implements the WordPress landing screen redesign behind the respective feature flag. The approach is to create a revamped fragment to co-exist with the current implementation, and use this new fragment conditionally (based on the flag) from the login activity. When the feature flag is turned on by default, the older flow can be removed.

This design is implemented via Jetpack Compose, ~~and a few things are not complete~~. To keep scope minimal for each PR, we can leverage the feature flag to develop partially complete UI with more granular feedback.

Since toggling the feature flag requires login, and navigating to this screen requires being logged out, it can be cumbersome to go back and forth between implementations. To make this easier, it is useful to apply the following patch:

<details>
<summary>Hard-coded flag patch</summary>

```diff
diff --git a/WordPress/src/main/java/org/wordpress/android/ui/accounts/LoginActivity.java b/WordPress/src/main/java/org/wordpress/android/ui/accounts/LoginActivity.java
index df7d849e4d..fccaf5aa14 100644
--- a/WordPress/src/main/java/org/wordpress/android/ui/accounts/LoginActivity.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/accounts/LoginActivity.java
@@ -227,7 +227,7 @@ public class LoginActivity extends LocaleAwareActivity implements ConnectionCall
     }
 
     private void loginFromPrologue() {
-        if (mLandingScreenRevampFeatureConfig.isEnabled()) {
+        if (true) {
             showFragment(new LoginPrologueRevampedFragment(), LoginPrologueRevampedFragment.TAG);
         } else {
             showFragment(new LoginPrologueFragment(), LoginPrologueFragment.TAG);
```
</details>

<details>

<summary>Former Solved Details</summary>

~~I also wired up one of the buttons to the login flow, because without the above hack, it isn't possible to toggle the feature flag off once logged out otherwise. The button functionality is scoped for a later PR, so this is an exception for practicality.~~

**Update:** Both login buttons were updated to have the expected functionality.

#### Questions

While this PR is still in draft, there are a couple of questions worth discussing before finalizing it:

1. The added font resource contributes significant size to the apk (~744k), but we only use this font in this one screen. Do we plan to use this font else-where, or is this considered acceptable?
    * There is a possibility to use "Downloadable fonts" for this, reducing the apk size, but this raises other questions, such as how to support users on older Android versions where this feature is not supported, etc.
2. The brush stroke image was rasterized directly from Figma as a 4x png. This seems to look ok in my testing, but it would be great to see how it looks on various form-factors. In the designs, the brush stroke is on the right side of the screen, but in this implementation, for RTL locales, it is on the left side of the screen. Is this ok?
    * Also, the image in RTL is not horizontally flipped. Since it is just a vertical brush stroke, this seemed ok to me, but it would be great to hear some other opinions.


**Note:** ~~The text for these buttons is based on future designs, and may possibly be reverted for the first iteration. I have deferred this decision for now.~~

Update: This has been reverted back to the former texts.

</details>

### Screenshots - Login Prologue

| Light | Dark |
|-|-|
|  ![wordpress-prologue](https://user-images.githubusercontent.com/8507675/187534368-a95ee014-ca4b-40f5-8727-4d08b52b6061.png) | ![wordpress-prologue-dark](https://user-images.githubusercontent.com/8507675/187534410-bd053fea-833d-42de-9ded-e433f0b67a83.png) |

| RTL | Landscape |
|-|-|
| ![wordpress-prologue-rtl](https://user-images.githubusercontent.com/8507675/187534656-e2a0b595-7585-4ebb-a864-501c1543d650.png) | ![wordpress-prologue-landscape](https://user-images.githubusercontent.com/8507675/187534674-91cc6a8c-b8d0-42e0-9e9b-a19946d5230a.png) |

To test:

Precondition: Enable the feature flag (either hard-coding via the patch above, or toggling within the app - or both)

1. Logout of the app once the feature flag is enabled
2. The new login screen should be visible
3. Put the device in dark mode
4. The dark mode design should be readable
5. Change to an RTL locale
6. Change the device orientation
7. Expect each view to be show as above

Login:

1. From the login screen, tap the "Log in or sign up with WordPress.com" button
2. Expect to be presented with the login flow
3. Go back
4. Tap the "Enter your existing site address" button
5. Expect to be presented with the site address login flow

Precondition: Disable the feature flag

1. Logout of the app
2. The old screen should be visible

Jetpack:

The Jetpack flavor of the app should be unaffected by these changes.

1. Build (or install from this PR) the Jetpack flavor of the app
2. Logout (with the feature enabled or disabled - it shouldn't matter)
6. The login screen should be unaffected

## Regression Notes
1. Potential unintended areas of impact
WordPress and Jetpack login screen

8. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing

9. What automated tests I added (or what prevented me from doing so)
This PR only adds UI changes (under a feature flag).

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Commit:
Rename package-name directories to resolve lint warning

Commit:
Add WordPress gridicon drawable

Note: it is possible that this will need to be optimized.

Commit:
Add button components for revamped login

Commit:
Add tagline components for revamped login

Commit:
Implement revamped login prologue fragment

Commit:
Show revamped login fragment behind feature flag

Commit:
Add brush stroke as raster image

This will likely need some modifications. Initially, I attempted to
import the vector image for the brush stroke, but the paths were too
long, and could not be parsed during compilation. I rasterize them in
this commit, but this may not work well in various screen sizes /
orientations.

Commit:
Split paths in WordPress gridicon icon

Commit:
Move WordPress gridicon icon to WordPress resource directory

mkevins committed:
Move brush stroke raster images to WordPress resource directory

mkevins committed:
Add font resource and adjust line height of tagline

Note: this adds a significant size to the apk (~744k), so it is possible
that we may opt to forgo this. If so, this branch can be rebased to
remove this from the repo history as well (instead of reverting).

ovitrif commented on WordPress/src/wordpress/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedFragment.kt:
```diff
@@ -0,0 +1,98 @@
+        const val TAG = "login_prologue_revamped_fragment_tag"
+    }
+
+    @Suppress("TooGenericExceptionThrown")
```
NP: I believe we can throw a different exception to get rid of this suppression.
I've been trying to throw an `IllegalStateException` in my WIP work on the JP landing screen revamp and then detekt suggested to use `error("message_here")`:

<img width="790" alt="Screenshot 2022-08-31 at 17 00 51" src="https://user-images.githubusercontent.com/4588074/187711497-3039f7fa-2024-4e82-bf06-76b725358ca9.png">

Now that I've played with `check` I tend believe it's even nicer than [`error()` inside an `if` statement](https://github.com/wordpress-mobile/WordPress-Android/compare/trunk...issue/17073-jetpack-landing-screen-revamp#diff-0ddbcac04d7f8e41dd4ecdc185e45a7b357addd9d9822210e14ab73c2267bccdR51-R53)

<img width="905" alt="Screenshot 2022-08-31 at 17 03 19" src="https://user-images.githubusercontent.com/4588074/187712131-997ffe99-b5dc-4d03-abb7-e4ba9c0b64ff.png">

Wdyt?

mkevins replied:
That sounds great Ovi! I hadn't given this much thought, so I'm glad you brought it up.

mkevins replied:
Addressed here: https://github.com/wordpress-mobile/WordPress-Android/pull/17104/commits/1ebcc6c17664ac5455a1ea44b81c7899c48cafc6

ovitrif commented on WordPress/src/wordpress/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedFragment.kt:
```diff
@@ -0,0 +1,98 @@
+        Image(
+                painter = painterResource(drawable.brush_stroke),
+                contentDescription = stringResource(string.login_prologue_revamped_content_description_wordpress_icon),
+                modifier = Modifier.align(Alignment.CenterEnd),
```
NP: One thing that I've learned so far in my Compose journey is that placing modifier argument last helps  to easily modify things later if changes after the modifier argument are needed.

Not sure if you agree but I wanted to share this :)

mkevins replied:
Good idea, I can move this to the end. 👍

mkevins replied:
I've addressed this (along with a few other similar instances) with this commit: https://github.com/wordpress-mobile/WordPress-Android/pull/17104/commits/a43ad1a0c89bc759702e06a129dfd33f232871d9

ovitrif commented on WordPress/src/wordpress/java/org/wordpress/android/ui/accounts/login/compose/components/Tagline.kt:
```diff
@@ -0,0 +1,72 @@
+                Modifier
+                        .width(234.dp)
+                        .padding(top = 25.dp),
+                fontFamily = FontFamily(Font(font.eb_garamond)),
```
This is looking good to me at the moment, though I have a feeling at some point in the near future we'll have to define a material typography theme.

mkevins replied:
Yeah, I have deferred hoisting this, since this is so far the only use, but it certainly makes sense to make this more "global" when we have other uses. 👍

mkevins replied:
Actually, I decided, since we seem to be planning more usage of this font, we may as well think forward to wrangle this here, since it's simple enough to do. I've extracted it in this commit: https://github.com/wordpress-mobile/WordPress-Android/pull/17104/commits/d41cad93366b719672c98ae1330cbc0fb39bc6f4

mkevins replied:
Actually, I decided, since we seem to be planning more usage of this font, we may as well think forward to wrangle this here, since it's simple enough to do. I've extracted it in this commit: https://github.com/wordpress-mobile/WordPress-Android/pull/17104/commits/d41cad93366b719672c98ae1330cbc0fb39bc6f4

ovitrif reviewed (commented):
Thank you for your work on this @mkevins.

It's lovely to see Compose work on the project, kudos for stepping forward, doing this and suggesting it for myself too as I'm working on the JP landing screen revamp 🙇 .

The code LGTM, I have a few comments that are part of the review but nothing big.

There's 3 more things I've been thinking of and wanted to share:

1. For some reason, when tapping the back button on the screen that comes after tapping the "Connect to your WordPress site" it gets me out of the app.
2. I was thinking if we could place the bg brush stroke in Landscape not to the `end` edge, but rather at some specific percentile offset. I like how it appears behind the other UI elements, but in landscape orientation it feels like it's just there somewhere in the background. Wdyt?
3. Are you planning to make the view "full screen"? It feels like it gives a more immersive experience imho when looking at the designs. Those UI bars are cutting stuff 😢 and removing that feeling a bit. My prefs goes for the original design with the full-screen experience:

<img width="297" alt="Screenshot 2022-08-31 at 17 18 32" src="https://user-images.githubusercontent.com/4588074/187715523-a47511b8-7885-4cd2-958d-86abb3f0d156.png">

mkevins commented:
> * For some reason, when tapping the back button on the screen that comes after tapping the "Connect to your WordPress site" it gets me out of the app.

This is actually the existing behavior.. I haven't looked into why this is happening, but if I can find the root cause, I'll try to see if it's in scope to address it in this revamp.

> * I was thinking if we could place the bg brush stroke in Landscape not to the `end` edge, but rather at some specific percentile offset. I like how it appears behind the other UI elements, but in landscape orientation it feels like it's just there somewhere in the background. Wdyt?

I'm not sure I entirely follow what you mean here.. but I guess you mean to move the brush stroke closer to the middle? It can certainly be achieved with some spacer and a percentage - this should be easy to change if so. I'll defer this until the question about the RTL brush stroke is resolved, since it may be something that can be tackled together.

> * Are you planning to make the view "full screen"? It feels like it gives a more immersive experience imho when looking at the designs. Those UI bars are cutting stuff 😢 and removing that feeling a bit. My prefs goes for the original design with the full-screen experience:

I'll try with fullscreen, good point!

mkevins committed:
Move modifier to last argument of composables

mkevins committed:
Use check to enforce LoginPrologueListener contract

mkevins committed:
Hoist EB Garamond font family to a more "global" place

ovitrif commented:
> I'm not sure I entirely follow what you mean here.. but I guess you mean to move the brush stroke closer to the middle? It can certainly be achieved with some spacer and a percentage - this should be easy to change if so. I'll defer this until the question about the RTL brush stroke is resolved, since it may be something that can be tackled together.

It was indeed what you're describing here 👍 
Good idea with deferring until we see what feedback we get on the RTL brush stroke 🙇

mkevins committed:
Set layout flag for no limits to make landing screen more immersive

mkevins committed:
Move WordPress font asset to WordPress app

This helps prevent bloating the Jetpack app with unused resources.

mkevins committed:
Make WordPress icon background opaque

This is to avoid having the brushstroke visible beneath the icon in
scenarios where it overlaps. This also sets a circle clip shape for the
icon so that the background color doesn't create a rectangular artifact
on top of the brush stroke background.

mkevins committed:
Update button style to match the latest Figma designs

mkevins committed:
Use drawBehind to draw brush stroke

Using a Canvas seems to be necessary here to avoid "pre-clipping" the
image to the container size, which would otherwise leave a small gap
when offsetting by a negative value on the y-axis.

mkevins committed:
Use a Pixel 3a for the preview dimensions

This seems better than using arbitrarily chosen dimensions.

ovitrif commented:
> > * For some reason, when tapping the back button on the screen that comes after tapping the "Connect to your WordPress site" it gets me out of the app.
> 
> This is actually the existing behavior.. I haven't looked into why this is happening, but if I can find the root cause, I'll try to see if it's in scope to address it in this revamp.

I found the root cause of this issue and added a fix on my PR. See comment: https://github.com/wordpress-mobile/WordPress-Android/pull/17117#issuecomment-1239243088

@mkevins Can you please add the same change to this PR, and perhaps target the feature branch [`feature/revamp-landing-screen`](https://github.com/wordpress-mobile/WordPress-Android/tree/) since my PR is supposed to be merged after yours according to its [Merging Strategy](https://github.com/wordpress-mobile/WordPress-Android/pull/17117#issue-1360306846).

Thanks in advance 🙇

mkevins committed:
Fix crash when pressing back button from login screen

mkevins commented:
> Can you please add the same change to this PR, and perhaps target the feature branch [`feature/revamp-landing-screen`](https://github.com/wordpress-mobile/WordPress-Android/tree/) since my PR is supposed to be merged after yours according to its [Merging Strategy](https://github.com/wordpress-mobile/WordPress-Android/pull/17117#issue-1360306846).
> 
> Thanks in advance

Good idea Ovi! I've changed the merge target, and added the crash fix to this PR as well. Also, I've updated the feature branch to be up to date with `trunk`. This is ready for another pass.

ovitrif reviewed (commented):
Great work @mkevins 👏 

Thank you for wrangling this 👍 

Everything works as described, I only found two small issues.

1. The appbar appears behind the transparent statusbar on the `Connect to your WordPress site` screen:
   <img width="314" src="https://user-images.githubusercontent.com/4588074/188942586-00c0d85c-e025-4fd8-93be-c10cbbaa4262.png">

   I presume the same fix that was done [here](https://github.com/wordpress-mobile/WordPress-Android/commit/836e631d7d19a9dd7aac2bbbd84388973b3e7e15) would work on this PR. That is moving the code that clears the `FLAG_LAYOUT_NO_LIMITS` flag to `onPause`.
More info in [this comment](https://github.com/wordpress-mobile/WordPress-Android/pull/17117#issuecomment-1239232637).

2. I noticed a rather unwanted effect when tapping the transparent `Looking to create a new site?` button.

Preview:

https://user-images.githubusercontent.com/4588074/188943183-c63241da-4b5b-4bd8-b512-a3e1b25ed314.mp4

ovitrif commented:
I'll be continuing on the great work started by Matthew to get this PR ready, as discussed via dm 🙇

ovitrif committed:
Fix restore from full screen on the login screen

antonis reviewed (commented):
I've tested on an older Huawei P9 (Android 7.0) and get the following error while starting up the WordPress app:
```
java.lang.IllegalStateException: Cannot create Typeface from ResourceFont(resId=2131296256, weight=FontWeight(weight=400),
```
([full stack trace](https://gist.github.com/antonis/ab8409b1d5b8ff6c18a6df74ed34b1c6)).
I'll investigate further and report back.

Edit:
[Commenting out the custom font assignment](https://github.com/wordpress-mobile/WordPress-Android/pull/17104/files#diff-063a391f634135b321990c648e7f74046a723c0ba9d6a68cb971b156d8743ba2R55) prevents the crash. My guess is that this is a memory or compose issue on older devices.

The brush does not load at all on the background though. I'm not sure yet if this is due to the screen size or the resource fails to load silently 🤔 
<img src="https://user-images.githubusercontent.com/304044/191292319-569582b6-9d78-45e3-ada2-717b0e9dd196.png" width=360 />

ovitrif committed:
Use legacy texts for login buttons

ovitrif committed:
Implement callback for site address login click

ovitrif committed:
Fix button shadow

ovitrif committed:
Lock LoginActivity to portrait mode

ovitrif committed:
Provide brush_stroke image in resource set for different screen densities

ovitrif committed:
Fallback to default Serif font if EBGaramond can't be loaded (to fix Huawei Android 7 issue)

antonis reviewed (approved):
Great work @mkevins and @ovitrif 👍 
I've retested with the latest changes both on my Pixel 5 (Android 13) and my Huawei P9 (Android 7) and [the issues reported before are fixed](https://github.com/wordpress-mobile/WordPress-Android/pull/17104#pullrequestreview-1113968784). The code LGTM too 🎉 
As discussed the CI check failures will be handled separately from another PR also pointing on `feature/revamp-landing-screen`

mkevins commented:
Thank you for addressing the remaining issues on this one Ovi! And thanks for testing and reviewing Antonis! 🎉
